src/components/ImageUpload/ImageUploadProvider.tsx

```diff
@@ -0,0 +1,273 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useWidgetApi } from '@matrix-widget-toolkit/react';
+import { TFunction } from 'i18next';
+import { PropsWithChildren, createContext, useCallback, useRef } from 'react';
+import { ErrorCode, FileRejection } from 'react-dropzone';
+import { useTranslation } from 'react-i18next';
+import { determineImageSize } from '../../lib';
+import { Size } from '../../state';
+import { PromiseSettledResult } from '../../types';
+import { SnackbarDismissAction, SnackbarProps, useSnackbar } from '../Snackbar';
+import { useMaxUploadSize } from './useMaxUploadSize';
+
+/**
+ * Maximum allowed resolution in pixels.
+ */
+const maxResolutionPixels = 32_000_000;
+
+type UploadErrors = {
+  /** Name of files that failed due to size restrictions */
+  size: string[];
+  /** Name of files that failed due to resolution restrictions */
+  resolution: string[];
+  /** Name of files that failed for other reasons */
+  other: string[];
+};
+
+type ImageUploadResult = {
+  fileName: string;
+  size: Size;
+  /**
+   * MXC URI pointing to the uploaded image.
+   * {@link https://spec.matrix.org/v1.9/client-server-api/#matrix-content-mxc-uris}
+   */
+  mxc: string;
+};
+
+export type ImageUploadContextValue = {
+  /**
+   * Handle a file drop from react-dropzone.
+   * Show a loading snackbar as long as uploads are in progress.
+   * Once all uploads are finished show an error snackbar for errors, if any.
+   * Can be called multiple times while other uploads are still running.
+   *
+   * @param files - Files to be uploaded from react-dropzone
+   * @param rejectedFiles - Files rejected by react-dropzone
+   * @returns Results for the uploaded files
+   */
+  handleDrop: (
+    files: File[],
+    rejectedFiles: FileRejection[],
+  ) => Promise<PromiseSettledResult<ImageUploadResult>[]>;
+  /**
+   * Maximum upload size allowed by the server in bytes.
+   */
+  maxUploadSizeBytes: number;
+};
+
+class ResolutionTooHighError extends Error {
+  constructor() {
+    super('Resolution too high');
+  }
+}
+
+export const ImageUploadContext = createContext<
+  ImageUploadContextValue | undefined
+>(undefined);
+
+export function ImageUploadProvider({ children }: PropsWithChildren<{}>) {
+  const { t } = useTranslation();
+  const { clearSnackbar, showSnackbar } = useSnackbar();
+  const widgetApi = useWidgetApi();
+  /** Holds all files that are to be uploaded */
+  const uploadQueue = useRef(new Set<File>());
+  const uploadErrors = useRef<UploadErrors>({
+    size: [],
+    resolution: [],
+    other: [],
+  });
+  const { maxUploadSizeBytes: serverMaxUploadSizeBytes } = useMaxUploadSize();
+  // limit image uploads to 25 MiB
+  const maxUploadSizeBytes = Math.min(serverMaxUploadSizeBytes, 26214400);
```

Comment:
what is the reasoning here? if the media repo doesn't set a limit, why should we? if really needed, why not a configurable env variable with a default value?

Reply:
Tests found that two things kill the board immediately: Large files and files with a high resolution. Both is now limited.

I would question to make it configurable, because then people will configure the board in a way that it doesn't really work (if you upload a large image). In most cases, the resolution limit should be reached anyway for large files.

Going to pull the const out and give it a comment why it is there.